### PR TITLE
Add CodeCompass client and server binaries and libraries.

### PR DESCRIPTION
**Pull request by gartung**
Everything except the webgui is built. This can probably be trimmed once the binaries needed to run the codecompass indexing are detemined.

### CODE REVIEW

**Comment by gartung** [comment]
@Dr15Jones @smuzaffar

**Comment by gartung** [comment]
@smuzaffar Is it possible to set PATH and LD_LIBRARY_PATH for an external package? odb and codecompass look for plugins in a relative path. I get these errors when I run using the symbolic links in external/slc7_amd64_gcc700/bin

[gartung@gartung]~% CodeCompass_parser 
terminate called after throwing an instance of 'std::runtime_error'
  what():  /home/gartung/../lib/parserplugin is not a directory!
zsh: abort (core dumped)  CodeCompass_parser
[gartung@gartung]~% odb
odb: error: no ODB plugin in '/home/gartung/CMSSW_10_2_0/external/slc7_amd64_gcc700/bin/../libexec/odb'

File: odb.spec [hunk 1]
```diff
@@ -0,0 +1,29 @@
+### RPM external odb 2.5.0-b.3
+Source0: https://codesynthesis.com/~boris/tmp/odb/pre-release/b.3/odb-2.5.0-b.3.tar.gz
```
Comment (davidlange6) [inline odb.spec]: this is a particularly interesting place to get source code from. What is this thing?
(added bonus, GPLv3. naively -1)

Comment (gartung) [reply odb.spec]: ODB is from codesynthesis. It is an object releational database. The odb command generates database code based on a model. CodeCompass uses this to interface with a database that it generates.

Comment (gartung) [reply odb.spec]: https://codesynthesis.com/products/odb/
The beta version is to support compiling with newer compilers that support c++11.

**Comment by davidlange6** [comment]
from someones tmp directory? Nice:)

**Comment by davidlange6** [comment]
could we instead pick up the production version

https://www.codesynthesis.com/download/odb/2.4/odb-2.4.0.tar.gz

File: odb.spec (hunk 1 above)
Comment (gartung) [reply odb.spec]: CodeCompass is something Chris asked to look into earlier this year.
https://github.com/Ericsson/CodeCompass/blob/master/doc/deps.md

**Comment by gartung** [comment]
I tried compiling odb 2.4.0 with gcc7 earlier this year but it fails because it makes use of std::auto_ptr. I did not realize that gcc7 defaults to std=c++14. I can try again.

**Comment by gartung** [comment]
“Some guy” is the primary author.

**Comment by davidlange6** [comment]
I guess you understand its not a reliable spot...

so when does the 1+ year old beta release become prod?

**Comment by gartung** [comment]
There is an actively built git repositort with the latest tagged beta

 

https://git.codesynthesis.com/cgit/odb/libodb/tree/?h=2.5.0-b.7

 

I just need to figure out how to download it.

**Comment by gartung** [comment]
@davidlange6 This is why I needed to use the 2.5.0 beta

* The action "build-external+odb+2.4.0-cms" was not completed successfully because Failed to build odb. Log file in /home/gartung/cms-build/BUILD/slc7_amd64_gcc700/external/odb/2.4.0-cms/log. Final lines of the log file:
/home/gartung/cms-build/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.1/plugin/include/wide-int.h:1148:17: error: 'STATIC_ASSERT' cannot be used as a function
/home/gartung/cms-build/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.1/plugin/include/wide-int.h: In instantiation of 'static wi::storage_ref wi::int_traits<generic_wide_int<T> >::decompose(long int*, unsigned int, const generic_wide_int<storage>&) [with storage = wide_int_ref_storage<true>]':
/home/gartung/cms-build/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.1/plugin/include/wide-int.h:976:48:   required from 'wide_int_ref_storage<SE>::wide_int_ref_storage(const T&, unsigned int) [with T = generic_wide_int<wide_int_ref_storage<true> >; bool SE = true]'
/home/gartung/cms-build/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.1/plugin/include/wide-int.h:753:26:   required from 'generic_wide_int<T>::generic_wide_int(const T&, unsigned int) [with T = generic_wide_int<wide_int_ref_storage<true> >; storage = wide_int_ref_storage<true>]'
/home/gartung/cms-build/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.1/plugin/include/wide-int.h:1827:25:   required from 'bool wi::ltu_p(const T1&, const T2&) [with T1 = generic_wide_int<wide_int_ref_storage<true> >; T2 = unsigned int]'
/home/gartung/cms-build/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.1/plugin/include/wide-int.h:1924:17:   required from 'bool wi::geu_p(const T1&, const T2&) [with T1 = generic_wide_int<wide_int_ref_storage<true> >; T2 = unsigned int]'
/home/gartung/cms-build/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.1/plugin/include/wide-int.h:2896:13:   required from 'typename wi::unary_traits<T>::result_type wi::lshift(const T1&, const T2&) [with T1 = generic_wide_int<wi::extended_tree<128> >; T2 = int; typename wi::unary_traits<T>::result_type = generic_wide_int<fixed_wide_int_storage<128> >]'
/home/gartung/cms-build/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.1/plugin/include/wide-int.h:3118:21:   required from 'typename wi::binary_traits<T1, T2>::signed_shift_result_type operator<<(const T1&, const T2&) [with T1 = generic_wide_int<wi::extended_tree<128> >; T2 = int; typename wi::binary_traits<T1, T2>::signed_shift_result_type = generic_wide_int<fixed_wide_int_storage<128> >]'
/home/gartung/cms-build/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.1/plugin/include/tree.h:5442:57:   required from here
/home/gartung/cms-build/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/bin/../lib/gcc/x86_64-unknown-linux-gnu/7.3.1/plugin/include/wide-int.h:928:23: error: 'gcc_checking_assert' was not declared in this scope
gcc_checking_assert (precision == x.get_precision ());
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[1]: *** [context.lo] Error 1
make[1]: Leaving directory `/home/gartung/cms-build/BUILD/slc7_amd64_gcc700/external/odb/2.4.0-cms/odb-2.4.0/odb'
make: *** [install-recursive] Error 1
error: Bad exit status from /home/gartung/cms-build/tmp/rpm-tmp.bkqPxq (%build)


RPM build errors:
Bad exit status from /home/gartung/cms-build/tmp/rpm-tmp.bkqPxq (%build)

**Comment by Dr15Jones** [comment]
If this package is requiring us to use beta code then I think we can say our evaluation of it has failed and we will try again after it has matured.

**Comment by gartung** [comment]
I was able to build the production version of odb (2.4.0) with gcc 6 patches from the fedora24 source rpm.

**Comment by gartung** [comment]
The CodeCompass_parser command uses compile_commands.json but is does not set an environment when it executes the compile command. COMPILER_RUNTIME_OBJECTS needs to be set so clang++ can find its header files. This may only work with a system installed llvm/clang. This may need to be installed in a fc28 docker image with gcc 7 and clang 6 to work.

In file included from /home/gartung/CMSSW_10_2_0/src/IOMC/RandomEngine/src/RandomEngineStateProducer.cc:2:
In file included from /home/gartung/CMSSW_10_2_0/src/IOMC/RandomEngine/src/RandomEngineStateProducer.h:21:
In file included from /home/gartung/CMSSW_10_2_0/src/FWCore/Framework/interface/global/EDProducer.h:24:
In file included from /home/gartung/CMSSW_10_2_0/src/FWCore/Framework/interface/global/producerAbilityToImplementor.h:25:
In file included from /home/gartung/CMSSW_10_2_0/src/FWCore/Framework/interface/global/implementors.h:22:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/memory:63:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/allocator.h:46:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/x86_64-unknown-linux-gnu/bits/c++allocator.h:33:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/ext/new_allocator.h:33:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/new:40:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/exception:142:
In file included from /cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/exception_ptr.h:38:
/cvmfs/cms.cern.ch/slc7_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/cxxabi_init_exception.h:38:10: fatal error: 'stddef.h' file not found
#include <stddef.h>

**Pull request closed without merging**